source/extensions/filters/http/ext_authz/ext_authz.cc

```diff
@@ -488,47 +488,43 @@ Http::FilterHeadersStatus Filter::encodeHeaders(Http::ResponseHeaderMap& headers
                    "set to the encoded response:",
                    *encoder_callbacks_, response_headers_to_add_.size(),
                    response_headers_to_set_.size());
-  bool omitted_response_headers = false;
   if (!response_headers_to_add_.empty()) {
     ENVOY_STREAM_LOG(
         trace, "ext_authz filter added header(s) to the encoded response:", *encoder_callbacks_);
     for (const auto& [key, value] : response_headers_to_add_) {
-      if (config_->enforceResponseHeaderLimits() && headers.size() >= headers.maxHeadersCount()) {
-        omitted_response_headers = true;
-        break;
-      }
       ENVOY_STREAM_LOG(trace, "'{}':'{}'", *encoder_callbacks_, key.get(), value);
       headers.addCopy(key, value);
+      if (config_->enforceResponseHeaderLimits() && !headersWithinLimits(headers)) {
+        responseHeaderLimitsReached();
+        return Http::FilterHeadersStatus::StopAllIterationAndWatermark;
+      }
     }
   }
 
   if (!response_headers_to_set_.empty()) {
     ENVOY_STREAM_LOG(
         trace, "ext_authz filter set header(s) to the encoded response:", *encoder_callbacks_);
     for (const auto& [key, value] : response_headers_to_set_) {
-      if (config_->enforceResponseHeaderLimits() && headers.get(key).empty() &&
-          headers.size() >= headers.maxHeadersCount()) {
-        omitted_response_headers = true;
-        // Continue because there could be other existing headers that can be set without increasing
-        // the number of headers.
-        continue;
-      }
       ENVOY_STREAM_LOG(trace, "'{}':'{}'", *encoder_callbacks_, key.get(), value);
       headers.setCopy(key, value);
+      if (config_->enforceResponseHeaderLimits() && !headersWithinLimits(headers)) {
+        responseHeaderLimitsReached();
+        return Http::FilterHeadersStatus::StopAllIterationAndWatermark;
+      }
     }
   }
 
   if (!response_headers_to_add_if_absent_.empty()) {
     for (const auto& [key, value] : response_headers_to_add_if_absent_) {
       ENVOY_STREAM_LOG(trace, "'{}':'{}'", *encoder_callbacks_, key.get(), value);
       if (auto header_entry = headers.get(key); header_entry.empty()) {
-        if (config_->enforceResponseHeaderLimits() && headers.size() >= headers.maxHeadersCount()) {
-          omitted_response_headers = true;
-          break;
-        }
         ENVOY_STREAM_LOG(trace, "ext_authz filter added header(s) to the encoded response:",
                          *encoder_callbacks_);
         headers.addCopy(key, value);
+        if (config_->enforceResponseHeaderLimits() && !headersWithinLimits(headers)) {
+          responseHeaderLimitsReached();
+          return Http::FilterHeadersStatus::StopAllIterationAndWatermark;
```

Comment:
FWIW,  Returning `StopIteration ` will also abort the filter chain iteration if localReply was sent 

```
  // Note that if a local reply was sent, no further iteration for headers as well as data and
  // trailers for the current filter and the filters following will happen. A local reply can be
  // triggered via sendLocalReply() or encodeHeaders().
 ```

I don't see a clear downside of returning `StopAllIterationAndWatermark` though, since filter chain should be aborted.  @ggreenway  wdyt?